transport/src/main/java/io/netty/channel/SendDirectByteBufPool.java

```diff
@@ -0,0 +1,1189 @@
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.internal.UnstableApi;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.ref.SoftReference;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+import java.nio.channels.GatheringByteChannel;
+import java.nio.channels.ScatteringByteChannel;
+import java.nio.charset.Charset;
+
+import static io.netty.util.internal.ObjectUtil.checkPositive;
+
+/**
+ * <strong>Internal usage only!</strong>
+ */
+@UnstableApi
+public final class SendDirectByteBufPool {
+
+    private static final int DEFAULT_PREALLOCATION_SIZE = 65536; // 64k
+    private PreallocationRef poolHead;
+    private Preallocation current = new Preallocation(DEFAULT_PREALLOCATION_SIZE);
+
+    /**
+     * Try to acquire a direct {@link ByteBuf} for the given size or return {@code null} if not possible.
+     *
+     * The call to {@link #acquire(int)} and {@link ByteBuf#release()} must be done in the same thread!
+     */
+    public ByteBuf acquire(int size) {
+        if (size > DEFAULT_PREALLOCATION_SIZE) {
+            return null;
+        }
+
+        Preallocation current = this.current;
+        ByteBuf buffer = current.buffer;
+        int remaining = buffer.writableBytes();
+
+        if (size < remaining) {
+            return newSendByteBuf(current, buffer, size);
+        }
+        if (size > remaining) {
+            this.current = current = getPreallocation();
+            buffer = current.buffer;
+            return newSendByteBuf(current, buffer, size);
+        }
+        // size == remaining
+        this.current = getPreallocation0();
+        return newSendByteBuf(current, buffer, size);
+    }
+
+    // Slice out a new buffer.
+    private PooledSendByteBuf newSendByteBuf(Preallocation current, ByteBuf buffer, int size) {
+        current.refCnt ++;
+        int writerIndex = buffer.writerIndex();
+        // Slice out a buffer and set reader and writerIndex to 0.
+        ByteBuf slice = buffer.slice(writerIndex, size).clear();
+
+        // Increase writerIndex of buffer from which we sliced out.
+        buffer.writerIndex(writerIndex + size);
+        return new PooledSendByteBuf(current, slice);
+    }
+
+    private Preallocation getPreallocation() {
+        Preallocation current = this.current;
+        if (current.refCnt == 0) {
+            current.buffer.clear();
+            return current;
+        }
+
+        return getPreallocation0();
+    }
+
+    private Preallocation getPreallocation0() {
+        PreallocationRef ref = poolHead;
+        if (ref != null) {
+            do {
+                Preallocation p = ref.get();
+                ref = ref.next;
+
+                if (p != null) {
+                    poolHead = ref;
+                    return p;
+                } else {
+                    ref.release();
+                }
+            } while (ref != null);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Change this condition so that it does not always evaluate to "true" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2583)